src/IO/S3Common.cpp

```diff
@@ -616,55 +617,86 @@ namespace S3
+                }
             }
-            if (name == S3)
+            else if (re2::RE2::PartialMatch(uri.getPath(), path_style_pattern, &bucket, &key))
             {
-                storage_name = name;
+                is_virtual_hosted_style = false;
+                endpoint = uri.getScheme() + "://" + uri.getAuthority();
             }
             else
-            {
-                storage_name = COSN;
-            }
+                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Bucket or key name are invalid in S3 URI.");
+
+            validateBucket(bucket);
+            validateKey(key);
         }
-        else if (re2::RE2::PartialMatch(uri.getPath(), path_style_pattern, &bucket, &key))
+        catch (const Exception & e)
         {
-            is_virtual_hosted_style = false;
-            endpoint = uri.getScheme() + "://" + uri.getAuthority();
-
-            /// S3 specification requires at least 3 and at most 63 characters in bucket name.
-            /// https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-s3-bucket-naming-requirements.html
-            if (bucket.length() < 3 || bucket.length() > 63)
-                throw Exception(ErrorCodes::BAD_ARGUMENTS, "Bucket name length is out of bounds in virtual hosted style S3 URI: {} ({})", quoteString(bucket), uri.toString());
+            throw Exception(e.code(), "{} ({})", e.message(), uri.toString());
         }
-        else
-            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Bucket or key name are invalid in S3 URI: {}", uri.toString());
+    }
+
+    void URI::validateBucket(const String & bucket)
+    {
+        /// See:
+        /// - https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
+        /// - https://cloud.ibm.com/apidocs/cos/cos-compatibility#createbucket
+
+        if (bucket.length() < 3 || bucket.length() > 222)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS,
+                "Bucket name length is out of bounds in virtual hosted style S3 URI: {}", quoteString(bucket));
+    }
+
+    void URI::validateKey(const String & /*key*/)
+    {
+        /// See:
+        /// - https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+        /// - https://cloud.ibm.com/apidocs/cos/cos-compatibility#putobject
+
+#if 0
+        /// The following is valid for AWS S3:
+        if (key.length() > 1024)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Key name is too long (showing first 1024 characters): {}", quoteString(key.substr(0, 1024) + "..."));
+        if (!ValidUTF8Impl::isValidUTF8(reinterpret_cast<const UInt8 *>(key.data()), key.size()))
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Key name must be valid UTF-8 string: {}", quoteString(key));
+#endif
+
+#if 0
+        /// The following is valid for IBM COS:
+        if (key.length() < 1)
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Key name is too short (0 bytes long)");
+#endif
```

Comment:
Need to be removed?

Reply:
I would leave it there as FIXME/documentation. I guess we may implement more strict checks at some point.

Reply:
But now we wont check that input is UTF8?

Reply:
Yes, because in COS non-UTF-8 is allowed. I guess in separate PR we may implement different checks depending on type of backend

Reply:
So, do we want to leave it without validation? I remember that we decided to validate utf8 + no ascii control chars + no glob chars + length. All is checked except utf8.

Reply:
Yes, because otherwise COS would be broken 🤷‍♂️